smart_campus/app/views.py

```diff
@@ -113,14 +119,145 @@ def index(request):
+                        os.remove(img.image.path)
+                    img.delete()
+
+                # Add new images (Max: 4 imgs)
+                for i in range(1, 5):
+                    if data['img{0}'.format(i)]:
+                        StationImage.objects.create(
+                            station=station,
+                            image=data['img{0}'.format(i)],
+                            is_primary=True if i == data['main_img_num'] else False
+                        )
+            station.save()
+
+            return HttpResponseRedirect('/stations/')
+
+        # if the datas failed in validation
+        # save the inputs back for redirection
+        form_data = form.cleaned_data
+    else:
+        form = StationForm()
+
+        # Load datas from the model instance
+        form_data = {
+            'name': station.name,
+            'category': station.category,
+            'content': station.content,
+            'beacon': station.beacon_set.first().name,
+            'lng': station.location.x,
+            'lat': station.location.y
+        }
+
+    if request.user.can(Permission.ADMIN):
+        beacon_set = Beacon.objects.all()
+    else:
+        beacon_set = Beacon.objects.filter(owner_group=request.user.group)
+
+    context = {
+        'email': request.user.email,
+        'categories': StationCategory.objects.all(),
+        'beacons': beacon_set,
+        'form': form,
+        'form_data': form_data,
+    }
+    return render(request, 'app/station_edit.html', context)
+
+
+@login_required
+def station_new_page(request):
+    if request.method == 'POST':
+        # Use custom StationForm to validate form datas
+        # is_valid() will be true if received data is good
+        # There'll be error messages in 'form' instance if validation failed
+        # cleaned_data will be the form inputs from the request
+        form = StationForm(request.POST, request.FILES)
+        if form.is_valid():
+            data = form.cleaned_data
+            station = form.save(commit=False)
+            station.location = Point(x=data['lng'], y=data['lat'])
+            station.owner_group = request.user.group
+            station.save()
+            # save first
+            # need an instance to add beacons
+            station.beacon_set.add(Beacon.objects.get(name=data['beacon']))
+            station.save()
+
+            # Add images (Max: 4 imgs)
+            # data['imgx'] will be None if not uploaded
+            for img_num in range(1, 5):
+                if data['img{0}'.format(img_num)]:
+                    StationImage.objects.create(
+                        station=station,
+                        image=data['img{0}'.format(img_num)],
+                        is_primary=True if img_num == data['main_img_num'] else False
+                    )
+            return HttpResponseRedirect('/stations/')
+    else:
+        form = StationForm()
+
+    if request.user.can(Permission.ADMIN):
```

Comment:
I suggest you rewrite as follow.

```python
if request.user.can(Permission.EDIT):
    beacon_set = Beacon.objects.filter(owner_group=request.user.group)
elif request.user.can(Permission.ADMIN):
    beacon_set = Beacon.objects.all()
else:
    some error
```

Reply:
Should I return  HttpResponseForbidden for that error?

Reply:
I think that can(Permission.ADMIN) should go first
With Permission ADMIN can EDIT

Reply:
You should return HttpResponseForbidden at the very beginning if one is not equipped with proper permission.

Reply:
Oh, I got your point.
I agree with that.